internal/services/sockets/handle_socket.go

```diff
@@ -0,0 +1,66 @@
+package sockets
+
+import (
+	"log/slog"
+	"net/http"
+	"sync"
+
+	"github.com/gorilla/websocket"
+)
+
+// Globals
+var connectedClients = NewConnectedClients()
+var streamChan = make(chan []byte, 100)
+var upgrader = websocket.Upgrader{
+	CheckOrigin: func(r *http.Request) bool {
+		return true //Allows all origins for now
+	},
+}
+
+// HandleSocket will accept websocket connections on '/ws/{clientID}'
+func HandleSocket(mux *http.ServeMux) {
+	mux.HandleFunc("GET /ws/{clientID}", func(w http.ResponseWriter, r *http.Request) {
+		clientID := r.PathValue("clientID")
+		var wg sync.WaitGroup
+		conn, err := upgrader.Upgrade(w, r, nil)
+		if err != nil {
+			slog.Error("Failed to upgrade connection", "reason", err)
+			http.Error(w, "Failed to upgrade connection", http.StatusInternalServerError)
+			return
+		}
+
+		if _, exists := connectedClients.GetClient(clientID); exists {
+			slog.Error("Client already connected", "reason", err)
+			return
+		}
+		connectedClients.AddClient(clientID, conn)
+		defer connectedClients.DeleteClient(clientID)
+
+		wg.Add(1)
+		//Remove client when they disconnect
+		go func() {
+			if conn, exists := connectedClients.GetClient(clientID); exists {
+				for {
+					_, _, err := conn.ReadMessage()
+					if err != nil {
+						connectedClients.DeleteClient(clientID)
+						wg.Done()
```

Comment:
Generally you should defer these when possible at the top of the go routine.

```go
go func() {
    defer wg.Done()
    panc("defers still runs to do clean up after panics")
}
```

Same thing with `Close()`, `Rollback()`, `Unlock()`, and other like functions, where we want to ensure cleanup happens. there are obviously exceptions, but generally its better practice to defer these.

The only time defers do not run is during `os.Exit()`, but this is because os.Exit is supposed to kill the application imidiately and ungracefully, wich is a feature not a bug.

Reply:
I should also add that using `panic()` is generally avoided. Instead exceptions are handled as `error`s. but sometimes we use code that produces a panic, or our application gets in a bad state due to erroneous assumptions about the code.

for example: 
- calling a method on a `nil` pointer
- accessing a slice out of bounds
- certain invalid math operations (eg divide by zero)

If a goroutine panics, and is not caught by  a `recover()`, then entire applicaiton will unwind and exit. 

this is why its good to use defers, because if the application panics, but recovers, defer function can save the application, leaving it in a clean enough state to keep running.